snapcraft/internal/meta/plugs.py

```diff
@@ -0,0 +1,134 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2019 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+
+from collections import OrderedDict
+from snapcraft.internal.meta.errors import PlugValidationError
+from typing import Any, Dict, List, Set  # noqa: F401
+
+logger = logging.getLogger(__name__)
+
+
+class Plug:
+    """Representation of a generic snap plug."""
+
+    def __init__(self, *, plug_name: str = "") -> None:
+        self._plug_dict = dict()  # type: Dict[str, Any]
+        self.plug_name = plug_name
+
+    def validate(self):
+        pass
+
+    @classmethod
+    def from_dict(cls, plug_dict: Dict[str, Any], plug_name: str = ""):
+        # If we explicitly support the type, use it instead.
+        if plug_dict.get("interface") == "content":
+            return ContentPlug.from_dict(plug_dict, plug_name=plug_name)
+
+        # Handle the general case.
+        plug = Plug(plug_name=plug_name)
+        plug._plug_dict = plug_dict
+        return plug
+
+    def to_yaml(self):
+        return OrderedDict(self._plug_dict)
+
+    def __repr__(self):
+        return repr(self.to_yaml())
+
+    def __str__(self):
+        return str(self.to_yaml())
+
+
+class ContentPlug(Plug):
+    """Representation of a snap content plug."""
+
+    def __init__(
+        self,
+        *,
+        plug_name: str = "",
+        content: str = "",
+        default_provider: str = "",
+        target: str = ""
+    ) -> None:
+        super().__init__(plug_name=plug_name)
+        self.interface = "content"  # type: str
```

Comment:
All plugs have an interface, why not pass it to the parent instead?
"content" could also be a enum like we have for build steps